dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/task/TaskSetState.java

```diff
@@ -16,60 +16,135 @@
+
+    private TaskResultSummary(String exception, Task<?> task, TaskState state) {
+      this.exception = exception;
+      this.task = task;
+      this.state = state;
+    }
+
+    @Nonnull
+    static TaskResultSummary create(Task<?> task, TaskResult<?> result) {
+      String optionalException = String.format(": %s", result.getException());
+      String exceptionDescription = result.getException() == null ? "" : optionalException;
+      return new TaskResultSummary(exceptionDescription, task, result.getState());
+    }
+
+    @Nonnull
+    @Override
+    public String toString() {
+      return String.format("Task %s (%s) %s%s", state, task.getCategory(), task, exception);
+    }
+  }
+
+  @ParametersAreNonnullByDefault
+  public static class TasksReport {
+
+    private final long count;
+    private final TaskState state;
+
+    @Nonnull
+    @Override
+    public String toString() {
+      return String.format("%d TASKS %s", count, state);
+    }
+
+    TasksReport(long count, TaskState state) {
+      this.count = count;
+      this.state = state;
+    }
+  }
+
   @ThreadSafe
   public static class Impl implements TaskSetState {
 
-    @GuardedBy("lock")
+    @GuardedBy("this")
     private final Map<Task<?>, TaskResult<?>> resultMap = new HashMap<>();
 
-    private final Object lock = new Object();
+    @Deprecated // Use TaskSetState instead of TaskSetState.Impl
+    @Override
+    public synchronized long failedRequiredTaskCount() {
+      long result = 0;
+      for (Task<?> key : resultMap.keySet()) {
+        TaskState state = resultMap.get(key).getState();
+        if (key.getCategory() == REQUIRED && state == FAILED) {
+          result++;
+        }
+      }
+      return result;
+    }
 
+    @Deprecated // Use TaskSetState instead of TaskSetState.Impl
     @Nonnull
-    public Map<Task<?>, TaskResult<?>> getTaskResultMap() {
-      synchronized (lock) {
-        return resultMap;
-      }
+    @Override
+    public synchronized ImmutableList<TaskResultSummary> taskResultSummaries() {
+      ImmutableList.Builder<TaskResultSummary> builder = ImmutableList.builder();
+      resultMap.forEach(
+          (task, result) -> {
+            TaskResultSummary summary = TaskResultSummary.create(task, result);
+            builder.add(summary);
+          });
+      return builder.build();
     }
 
+    @Deprecated // Use TaskSetState instead of TaskSetState.Impl
+    @Nonnull
     @Override
-    @SuppressWarnings("unchecked")
-    public <T> TaskResult<T> getTaskResult(@Nonnull Task<T> task) {
-      synchronized (lock) {
-        return (TaskResult<T>) resultMap.get(task);
-      }
+    public synchronized ImmutableList<TasksReport> tasksReports() {
+      ImmutableList.Builder<TasksReport> builder = ImmutableList.builder();
+      resultMap.values().stream()
+          .collect(Collectors.groupingBy(TaskResult::getState, Collectors.counting()))
+          .forEach((key, value) -> builder.add(new TasksReport(value, key)));
```

Comment:
I would recommend creating a second stream and collecting to list instead of using the builder.

```
.entrySet().stream().map(entry -> new TaskReport(...)).collect(toImmutableList());
```

Reply:
Done